File: apps/tui/src/lib/protocol.ts
import { z } from 'zod';

// ── Shared ────────────────────────────────────────────────────────────────────

export const IndexedQuerySchema = z.object({
  id: z.string(),
  filePath: z.string(),
  lineStart: z.number().int(),
  lineEnd: z.number().int(),
  functionName: z.string().nullable(),
  queryKind: z.enum(['TaggedTemplate', 'SqlFile', 'RawString']),
  rawSql: z.string(),
});
export type IndexedQuery = z.infer<typeof IndexedQuerySchema>;

// ── Requests (TS → Rust) ──────────────────────────────────────────────────────

export const RequestSchema = z.discriminatedUnion('type', [
  z.object({ type: z.literal('Ping'), id: z.string() }),
  z.object({ type: z.literal('ScanProject'), id: z.string(), rootDir: z.string() }),
  z.object({
    type: z.literal('RunQuery'),
    id: z.string(),
    sql: z.string(),
    connectionString: z.string(),
  }),
  z.object({
    type: z.literal('ExplainQuery'),
    id: z.string(),
    sql: z.string(),
    connectionString: z.string(),
    analyze: z.boolean(),
  }),
]);
export type Request = z.infer<typeof RequestSchema>;

// ── Responses (Rust → TS) ─────────────────────────────────────────────────────

export const ResponseSchema = z.discriminatedUnion('type', [
  z.object({ type: z.literal('Pong'), id: z.string() }),
  z.object({
    type: z.literal('ScannedProject'),
    id: z.string(),
    queries: z.array(IndexedQuerySchema),
  }),
  z.object({
    type: z.literal('QueryResult'),
    id: z.string(),
    columns: z.array(z.string()),
    rows: z.array(z.array(z.string().nullable())),
    rowCount: z.number(),
    durationMs: z.number(),
  }),
  z.object({ type: z.literal('ExplainResult'), id: z.string(), output: z.string() }),
  z.object({ type: z.literal('Error'), id: z.string(), message: z.string() }),
]);
export type Response = z.infer<typeof ResponseSchema>;

// ── Convenience narrow types ──────────────────────────────────────────────────

export type QueryResult = Extract<Response, { type: 'QueryResult' }>;
export type ExplainResult = Extract<Response, { type: 'ExplainResult' }>;

File: crates/engine/src/protocol.rs
use serde::{Deserialize, Serialize};

// ── IndexedQuery ──────────────────────────────────────────────────────────────

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct IndexedQuery {
    pub id: String,
    pub file_path: String,
    pub line_start: u32,
    pub line_end: u32,
    pub function_name: Option<String>,
    pub query_kind: QueryKind,
    pub raw_sql: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum QueryKind {
    TaggedTemplate,
    SqlFile,
    RawString,
}

// ── Requests (TS → Rust) ──────────────────────────────────────────────────────

#[derive(Debug, Deserialize)]
#[serde(tag = "type")]
pub enum Request {
    Ping {
        id: String,
    },
    ScanProject {
        id: String,
        #[serde(rename = "rootDir")]
        root_dir: String,
    },
    RunQuery {
        id: String,
        sql: String,
        #[serde(rename = "connectionString")]
        connection_string: String,
    },
    ExplainQuery {
        id: String,
        sql: String,
        #[serde(rename = "connectionString")]
        connection_string: String,
        analyze: bool,
    },
}

// ── Responses (Rust → TS) ─────────────────────────────────────────────────────

#[derive(Debug, Serialize)]
#[serde(tag = "type")]
pub enum Response {
    Pong {
        id: String,
    },
    ScannedProject {
        id: String,
        queries: Vec<IndexedQuery>,
    },
    QueryResult {
        id: String,
        columns: Vec<String>,
        rows: Vec<Vec<Option<String>>>,
        #[serde(rename = "rowCount")]
        row_count: i64,
        #[serde(rename = "durationMs")]
        duration_ms: f64,
    },
    ExplainResult {
        id: String,
        output: String,
    },
    Error {
        id: String,
        message: String,
    },
}
